apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessmentQuestion/assessmentQuestion.ts

```diff
@@ -136,6 +143,179 @@ router.post(
       } else {
         res.send({});
       }
+    } else if (req.body.__action === 'bot_grade_assessment') {
+      // check if bot grading is enabled
+      if (!res.locals.bot_grading_enabled) {
+        throw new error.HttpStatusError(403, 'Access denied (feature not available)');
+      }
+      console.log('Bot grading the assessment question');
+
+      if (config.openAiApiKey === null || config.openAiOrganization === null) {
+        throw new error.HttpStatusError(501, 'Not implemented (feature not available)');
+      }
+      const openaiconfig = {
+        apiKey: config.openAiApiKey,
+        organization: config.openAiOrganization,
+      };
+
+      const { urlPrefix, assessment, assessment_instance, assessment_question, authz_result } =
+        res.locals;
+      const question = res.locals.question;
+      const question_course = await getQuestionCourse(question, res.locals.course);
+
+      const serverJob = await createServerJob({
+        courseId: res.locals.course ? res.locals.course.id : null,
+        type: 'botGrading',
+        description: 'Use LLM to grade assessment question',
+      });
+
+      serverJob.executeInBackground(async (job) => {
+        const openai = new OpenAI(openaiconfig);
+
+        // get all instance questions
+        const result = await queryRows(
+          sql.select_instance_questions_manual_grading,
+          {
+            assessment_id: res.locals.assessment.id,
+            assessment_question_id: res.locals.assessment_question.id,
+          },
+          InstanceQuestionSchema,
+        );
```

Comment:
I'd encourage you to use a custom query that includes all the information you need in a single query, instead of separate queries for submission and variant. Also note that this query includes a lot of fields you don't need.

Reply:
I'm trying to return a submission and the variant at the same time. Currently I'll need both the Submission object and the Variant object as it's required by the questionModule.render function. I'm wondering how I should define the Schema so the return object type is correct?

Reply:
You can use `to_jsonb` to return the entire row as a query column. As an example, check:
https://github.com/PrairieLearn/PrairieLearn/blob/e660d36c2738b3ceabafbca381354fbc35a5e58d/apps/prairielearn/src/lib/externalGrader.sql#L1-L15
https://github.com/PrairieLearn/PrairieLearn/blob/e660d36c2738b3ceabafbca381354fbc35a5e58d/apps/prairielearn/src/lib/externalGrader.ts#L33-L39
https://github.com/PrairieLearn/PrairieLearn/blob/e660d36c2738b3ceabafbca381354fbc35a5e58d/apps/prairielearn/src/lib/externalGrader.ts#L67-L71

Reply:
That is exactly what I need! I was so close to getting everything earlier but the to_jsonb function!